homeassistant/components/group/sensor.py

```diff
@@ -665,19 +675,40 @@ def _get_valid_units(self) -> set[str | None]:
 
         If device class is set and compatible unit of measurements.
         If device class is not set, use one unit of measurement.
+        Only calculate valid units if there are no valid units set.
         """
-        if (
-            device_class := self.device_class
-        ) in UNIT_CONVERTERS and self.native_unit_of_measurement:
+        # If we once have a set of valid units we should not recalculate it
+        # when all states needs to be considered
+        if (valid_units := self._valid_units) and not self._ignore_non_numeric:
+            return valid_units
+
+        native_uom = self.native_unit_of_measurement
+        if (device_class := self.device_class) in UNIT_CONVERTERS and native_uom:
             self._can_convert = True
             return UNIT_CONVERTERS[device_class].VALID_UNITS
-        if (
-            device_class
-            and (device_class) in DEVICE_CLASS_UNITS
-            and self.native_unit_of_measurement
-        ):
+        if device_class and (device_class) in DEVICE_CLASS_UNITS and native_uom:
             valid_uoms: set = DEVICE_CLASS_UNITS[device_class]
             return valid_uoms
-        if device_class is None and self.native_unit_of_measurement:
-            return {self.native_unit_of_measurement}
+        if device_class is None and native_uom:
+            return {native_uom}
         return set()
+
+    def _get_state_of_entities(
+        self,
+    ) -> list[str]:
+        """Return if states are valid."""
+        state_of_entities: dict[str, bool] = {}
+        for entity_id in self._entity_ids:
+            if (state := self.hass.states.get(entity_id)) is not None:
+                try:
+                    float(state.state)
+                    state_of_entities[entity_id] = True
+                except ValueError:
+                    state_of_entities[entity_id] = False
+            else:
+                state_of_entities[entity_id] = False
+        return [
+            entity_id
+            for entity_id, state_valid in state_of_entities.items()
+            if state_valid is True
+        ]
```

Comment:
It doesn't seem like we need the intermediate dict?
Maybe we can simplify to something like:
```suggestion
        valid_entities: list[str] = []
        for entity_id in self._entity_ids:
            if (state := self.hass.states.get(entity_id)) is not None:
                with suppress(ValueError):
                    float(state.state)
                    valid_entities.append(entity_id)
        return valid_entities
```

Or with a utility function:
```py
    return [entity_id for entity_id in self._entity_ids if _has_numeric_state(hass, entity_id)]
```